### PR TITLE
Add driver for GW Instek PSP series

### PR DESCRIPTION
**Pull request by ettom**
The protocol is described in the manual, available [here](https://www.gwinstek.com/en-global/products/downloadSeriesDownNew/7327/1426).

Generally, everything seems to work well. One annoying limitation of the protocol/device is that it uses the same field for reporting both the actual voltage and the target voltage. This means that querying the target voltage is not possible while the output is enabled or the device is in CC mode. I did implement a heuristic to work around this as best as I could, but one corner case remains - when SmuView is launched while the output is enabled and the device is in CC mode, the target voltage will be reported as 0.00V until the output is turned on (at which point the target voltage will be updated according to the one set on the device), or until a new target voltage is set via the UI. I think that this limitation is acceptable if properly documented.

### CODE REVIEW

**Review by biot — CHANGES_REQUESTED** [review]

**Comment by biot** [comment]
Sorry about the long delay in getting this reviewed, I hope you still want to get this driver merged. The code looks good, aside from the minor changes. Seems like a challenging device to write a driver for.

File: src/hardware/gwinstek-psp/api.c [hunk 1]
```diff
@@ -17,125 +17,344 @@
...
+		}
+	}
+
+	if (!conn)
+		return NULL;
+	if (!serialcomm)
+		serialcomm = "2400/8n1";
+	if (!force_detect) {
+		sr_err("The gwinstek-psp driver requires the force_detect parameter");
+		return NULL;
```
[the remark this answers is not in the document]
Comment (ettom) [reply src/hardware/gwinstek-psp/api.c:128]: As far as I can tell, the only way would be to set the voltage or current limit to the maximum value (there are dedicated commands for this) and infer the model from that. Unfortunately, I couldn't get the vendor-provided software to work at all, so I can't verify if this is what it does. In any case, I think this method would be unacceptable, as the software shouldn't change the output values without explicit user input.

**Review by biot — CHANGES_REQUESTED** [review]

File: src/hardware/gwinstek-psp/protocol.c [hunk 2]
```diff
@@ -17,15 +17,189 @@
...
+ * Must be called right after calling gwinstek_psp_get_all_values.
+ */
+SR_PRIV int gwinstek_psp_check_terminator(struct sr_serial_dev_inst *serial,
+    struct dev_context *devc)
+{
+	int bytes_left, ret;
+
+	g_mutex_lock(&devc->rw_mutex);
+	/* Sleep for a while to be extra sure the device has sent everything */
+	g_usleep(500 * 1000);
```
[the remark this answers is not in the document]
Comment (ettom) [reply src/hardware/gwinstek-psp/protocol.c]: Good point.

I'm not completely sure why, but with 15ms, I could pretty reliably cause the wrong terminator to be detected - device is in URPSP1, but the driver would report URPSP2. I'm guessing the device might not be transmitting at a constant rate or has some delays? With a 17+ms delay, it was solid. So I've set it to 20ms, for just a bit of safety net.

**Review by biot — APPROVED** [review]

**Comment by abraxa** [comment]
Merged as https://sigrok.org/gitaction/libsigrok.git/bcf21cc..e938e2d, thank you for your contribution!

**Pull request closed without merging**